File: C++/divisible-game.cpp
// Time:  precompute: O(r)
//        runtime:    O(nlogx + f) = O(nlogx), f = sum of distinct prime factors over all nums (<= 7n)
// Space: O(r + n)

// number theory, kadane's algorithm, prefix sum
vector<int> linear_sieve_of_eratosthenes(int n) {  // Time: O(n), Space: O(n)
    vector<int> spf(n + 1, -1);
    vector<int> primes;
    for (int i = 2; i <= n; ++i) {
        if (spf[i] == -1) {
            spf[i] = i;
            primes.emplace_back(i);
        }
        for (const auto& p : primes) {
            if (i * p > n || p > spf[i]) {
                break;
            }
            spf[i * p] = p;
        }
    }
    return spf;
}

const int MAX_NUMS = 1e6;
const auto& SPF = linear_sieve_of_eratosthenes(MAX_NUMS);
class Solution {
public:
    int divisibleGame(vector<int>& nums) {
        static const int MOD = 1e9 + 7;

        vector<int> prefix(size(nums) + 1);
        for (int i = 0; i < size(nums); ++i) {
            prefix[i + 1] = prefix[i] + nums[i];
        }
        unordered_map<int, vector<int>> lookup;
        for (int i = 0; i < size(nums); ++i) {
            int x = nums[i];
            while (x != 1) {
                const auto& p = SPF[x];
                lookup[p].emplace_back(i);
                while (x % p == 0) {
                    x /= p;
                }
            }
        }
        int best_diff = -ranges::min(nums), best_k = 2;
        for (const auto& [p, idxs] : lookup) {
            int total = 0, j = -1;
            for (const auto& i : idxs) {
                total = max(total - (j != -1 ? prefix[(i - 1) + 1] - prefix[j + 1] : 0), 0) + nums[i];
                if (total > best_diff) {
                    best_diff = total;
                    best_k = p;
                } else if (total == best_diff) {
                    best_k = min(best_k, p);
                }
                j = i;
            }
        }
        return ((static_cast<int64_t>(best_diff) * best_k) % MOD + MOD) % MOD;
    }
};

File: Python/divisible-game.py
# Time:  precompute: O(r)
#        runtime:    O(nlogx + f) = O(nlogx), f = sum of distinct prime factors over all nums (<= 7n)
# Space: O(r + n)

import collections


# number theory, kadane's algorithm, prefix sum
def linear_sieve_of_eratosthenes(n):  # Time: O(n), Space: O(n)
    primes = []
    spf = [-1]*(n+1)  # the smallest prime factor
    for i in xrange(2, n+1):
        if spf[i] == -1:
            spf[i] = i
            primes.append(i)
        for p in primes:
            if i*p > n or p > spf[i]:
                break
            spf[i*p] = p
    return spf


MAX_NUMS = 10**6
SPF = linear_sieve_of_eratosthenes(MAX_NUMS)
class Solution(object):
    def divisibleGame(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        MOD = 10**9+7
        prefix = [0]*(len(nums)+1)
        for i in xrange(len(nums)):
            prefix[i+1] = prefix[i]+nums[i]
        lookup = collections.defaultdict(list)
        for i, x in enumerate(nums):
            while x != 1:
                p = SPF[x]
                lookup[p].append(i)
                while x%p == 0:
                    x //= p
        best_diff, best_k = -min(nums), 2
        for p, idxs in lookup.iteritems():
            total, j = 0, -1
            for i in idxs:
                total = max(total-(prefix[(i-1)+1]-prefix[j+1]), 0)+nums[i]
                if total > best_diff:
                    best_diff, best_k = total, p
                elif total == best_diff:
                    best_k = min(best_k, p)
                j = i
        return (best_diff*best_k)%MOD
